torch/csrc/jit/passes/quantization.cpp

```diff
@@ -676,5 +677,255 @@ TORCH_API script::Module InsertObservers(
+        // Observer forward call node
+        nodes_to_destroy.push_back(user);
+        // GetAttr node for observer module
+        nodes_to_destroy.push_back(user->inputs()[0]->node());
+      }
+    }
+
+    // calculate qparams and insert quant/dequant calls
+    auto calculate_qparams = (*observer_module).get_method("calculate_qparams");
+    IValue qparams = calculate_qparams(std::vector<IValue>());
+    Node* dequant;
+    if (v->node()->kind() == prim::GetAttr && v->node()->s(c10::attr::name) == "weight") {
+      dequant = insertQuantDeQuantCall(v, qparams, at::kQInt8);
+    } else {
+      dequant = insertQuantDeQuantCall(v, qparams, at::kQUInt8, insert_after);
+    }
+    v->replaceAllUsesWith(dequant->output());
+    Node* q = traverseToQuantNode(dequant);
+    TORCH_INTERNAL_ASSERT(q);
+    q->replaceInputWith(dequant->output(), v);
+  }
+}
+
+script::Module InsertQuantDeQuant(
+    const script::Module& module,
+    const std::string& method_name) {
+  // TODO: Uncomment
+  //script::Module module = input_module.clone();
+  script::Method method = module.get_method(method_name);
+  auto graph = method.graph();
+  TORCH_CHECK(graph != nullptr);
+  auto num_activation_inputs = method.num_inputs();
+  std::vector<Value*> values_to_observe;
+  std::vector<Value*> input_values;
+
+  // For traversing all blocks in the graph including subblocks.
+  std::stack<Block*> blocks_to_visit;
+
+  // Mark observer nodes for inputs so we dont add observers
+  // for observers while traversing graph
+  std::unordered_set<Node*> observer_for_input;
+
+  // Add observer for external input nodes excluding parameters
+  // These are treated as activation as they vary across batches
+  // and need to be observed.
+
+  // prim::Param nodes do not belong to the graph. Hence the Insert
+  // point is the beginning of graph node. This also safe guards against
+  // observing a potentially mutated value due to some in-place operation
+  for (size_t idx = 0; idx < num_activation_inputs; ++idx) {
+    auto& v = graph->inputs()[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      input_values.push_back(v);
+    }
+  }
+
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      // Skip nodes that we don't need to observe, e.g. 'prim::Constant' or
+      // observer nodes
+      if (!outputsNeedToBeObserved(n) || observer_for_input.count(n) != 0) {
+        continue;
+      }
+
+      // Record all outputs in the values_to_observe - we'll later add observers
+      // for all values from it.
+      for (Value* v : n->outputs()) {
+        values_to_observe.push_back(v);
+      }
+
+      // Schedule subblocks (if any) for visiting.
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+    }
+  }
+
+  // quantize bias first
+  for (Value* v : values_to_observe) {
```

Comment:
Identifying the bias nodes based on iterating the observable nodes and checking whether there is actually not observer and then checking for bias seems indirect approach.

Is there a different way possible. Can this not be a separate pass as it was earlier [quant-dequant for params]?

Reply:
don't we need to do the same in a separate pass?

Reply:
why would we like to split into multiple passes?

Reply:
I think the no-observer should actually be an assert when we find that the bias is an argument for "aten::conv", see "quantizeBias", we can probably move this inside that loop

Reply:
i think approach is different now. Earlier parameters did not have observers and so passes were different. Ok to have it done all together. 

Only mechanism to how we get to node is bit indirect.  Since bias should not have observer, we are looking at observable nodes and then checking if no observer module and if its bias seemed indirect to me.

Since bias is unrelated to observers, you can access directly by looking at params to module and mapping them to supported op argument [conv or linear]

Reply:
yeah makes sense, I just make a change to move this check inside quantizeBias

Reply:
we can certainly start from conv/linear and quantize the bias, but I think the current approach is fine too, please take a look again.